src/Domain/CoordinateMaps/Frustum.cpp

```diff
@@ -0,0 +1,339 @@
+  using ReturnType = std::decay_t<tt::remove_reference_wrapper_t<T>>;
+
+  const ReturnType& xi = source_coords[0];
+  const ReturnType& eta = source_coords[1];
+  const ReturnType& zeta = source_coords[2];
+
+  ReturnType physical_x =
+      0.25 * ((x1_ * (1.0 - xi) + x2_ * (1.0 + xi)) * (1.0 - zeta) +
+              (x3_ * (1.0 - xi) + x4_ * (1.0 + xi)) * (1.0 + zeta));
+  ReturnType physical_y =
+      0.25 * ((y1_ * (1.0 - eta) + y2_ * (1.0 + eta)) * (1.0 - zeta) +
+              (y3_ * (1.0 - eta) + y4_ * (1.0 + eta)) * (1.0 + zeta));
+  ReturnType physical_z =
+      0.5 * (lower_bound_ * (1.0 - zeta) + upper_bound_ * (1.0 + zeta));
+
+  // Assigns location of frustum based on passed "direction_of_frustum"
+  // Frustums on z axis:
+  if (direction_of_frustum_.axis() == Direction<3>::Axis::Zeta) {
+    return (direction_of_frustum_.side() == Side::Upper)
+               ? std::array<ReturnType, 3>{{std::move(physical_x),
+                                            std::move(physical_y),
+                                            std::move(physical_z)}}
+               : std::array<ReturnType, 3>{{std::move(physical_x),
+                                            std::move(-physical_y),
+                                            std::move(-physical_z)}};
+  }
+  // Frustums on y axis:
+  if (direction_of_frustum_.axis() == Direction<3>::Axis::Eta) {
+    return (direction_of_frustum_.side() == Side::Upper)
+               ? std::array<ReturnType, 3>{{std::move(physical_y),
+                                            std::move(physical_z),
+                                            std::move(physical_x)}}
+               : std::array<ReturnType, 3>{{std::move(physical_y),
+                                            std::move(-physical_z),
+                                            std::move(-physical_x)}};
+  }
+  // Frustums on x axis:
+  if (direction_of_frustum_.axis() == Direction<3>::Axis::Xi) {
+    return (direction_of_frustum_.side() == Side::Upper)
+               ? std::array<ReturnType, 3>{{std::move(physical_z),
+                                            std::move(physical_x),
+                                            std::move(physical_y)}}
+               : std::array<ReturnType, 3>{{std::move(-physical_z),
+                                            std::move(-physical_x),
+                                            std::move(physical_y)}};
+  }
+  ERROR("Improper axis passed to function.");
+}
+
+template <typename T>
+std::array<std::decay_t<tt::remove_reference_wrapper_t<T>>, 3> Frustum::inverse(
+    const std::array<T, 3>& target_coords) const noexcept {
+  using ReturnType = std::decay_t<tt::remove_reference_wrapper_t<T>>;
+
+  ReturnType physical_x;
+  ReturnType physical_y;
+  ReturnType physical_z;
+
+  // Assigns location of frustum based on passed "direction_of_frustum"
+  // Frustums on z axis:
+  switch (direction_of_frustum_.axis()) {
```

Comment:
See how BulgedCube is implemented with a lambda and use that approach when you factor things to get rid of the extra copies